prediction_market_agent/agents/microchain_agent/memory_functions.py

```diff
@@ -94,6 +97,44 @@ def __call__(self, context: str) -> str:
         return memories_to_learnings(memories=results, model=self.model)
 
 
+def fetch_memories_from_last_run(
+    agent_identifiers: list[AgentIdentifier],
+) -> dict[str, list[LongTermMemories]]:
+    """
+    Returns memories from the last run for each agent.
+    The last run is identified by looking at "GameRoundEnd" function calls
+    as part of the prompt history of each agent.
+    """
+    entries_from_latest_run: dict[str, list[LongTermMemories]] = {}
+    for agent_id in agent_identifiers:
+        ltm = LongTermMemoryTableHandler.from_agent_identifier(agent_id)
+        entries_for_agent = ltm.search()
+        # We need the 2nd GameRoundEnd. If there is only one, then start_date should be None.
+        game_round_occurrences = [
+            i
+            for i in entries_for_agent
+            if i.metadata_dict is not None and "content" in i.metadata_dict
+            and "GameRoundEnd(" in i.metadata_dict["content"]
+        ]
+        # We initially assume all memories should be processed (it's the case if we have # of game_round_occurrences
+        # <= 1). If game_round_occurrences > 1, at least 1 run was completed, hence we fetch the memories from the
```

Comment:
If `game_round_occurrences == 1` doesn't that also mean that 1 round was completed?

GameRoundEnd is called after the round finishes.

Reply:
Maybe my comment is not clear
-> If `game_round_end_count` == 0, no rounds (but should not occur) -> still needs to be implemented
-> If `game_round_end_count` == 1, 1 round, process all memories (current implementation OK)
> If `game_round_end_count` > 1, 1 round, process memories from 2nd game round end until the most recent one (current implementation OK)

So I would change only the first item above - would you agree?

Reply:
> If game_round_end_count > 1, 1 round, process memories from 2nd game round end

Not 2nd GameRoundEnd, because there could be 10 games already. Always taking everything from 2nd would get multiple games.

Shouldn't it be

1. If no `GameRoundEnd` is found, simply take all.
2. If `GameRoundEnd` is the latest (or technically 2nd to latest) message in the history, ignore it.
3. Take history from the latest `GameRoundEnd` onwards (unless it's the latest one from (2).

?

Reply:
Indeed! My bad. I didn't realize it's sorted DESC.

If this logic is dependent on hidden behavior that could change, maybe this is a good idea (although it won't do anything atm), but up to you if you commit it.

```suggestion
        # <= 1). If game_round_occurrences > 1, at least 1 run was completed, hence we fetch the memories from the
        game_round_occurrences = sorted(game_round_occurrences, key=lambda x: x.datetime_, reverse=True)
```

Reply:
done